core/src/main/kotlin/com/malinskiy/marathon/execution/progress/PoolProgressAccumulator.kt

```diff
@@ -0,0 +1,426 @@
+    }
+
+    private fun trackTestTransition(poolId: DevicePoolId, transition: StateMachine.Transition<TestState, TestEvent, TestAction>) {
+        val validTransition = transition as? StateMachine.Transition.Valid
+        val final = if (validTransition is StateMachine.Transition.Valid) {
+            when (validTransition.sideEffect) {
+                is TestAction.Conclude -> true
+                else -> false
+            }
+        } else false
+
+        val (testResult: TestResult?, device: DeviceInfo?) = extractEventAndDevice(transition)
+        if (testResult == null || device == null) return
+
+        track.test(poolId, device, testResult, final)
+    }
+
+    private fun extractEventAndDevice(transition: StateMachine.Transition<TestState, TestEvent, TestAction>): Pair<TestResult?, DeviceInfo?> {
+        val event = transition.event
+        val testResult: TestResult? = when (event) {
+            is TestEvent.Passed -> event.testResult
+            is TestEvent.Failed -> event.testResult
+            is TestEvent.Incomplete -> {
+                if (event.final) {
+                    event.testResult.copy(status = TestStatus.FAILURE)
+                } else {
+                    event.testResult
+                }
+            }
+            else -> null
+        }
+        val device: DeviceInfo? = when (event) {
+            is TestEvent.Passed -> event.device
+            is TestEvent.Failed -> event.device
+            is TestEvent.Incomplete -> event.device
+            else -> null
+        }
+        return Pair(testResult, device)
+    }
+
+    fun aggregateResult(): Boolean {
+        synchronized(tests) {
+            return tests.map {
+                when (it.value.state) {
+                    is TestState.Added -> {
+                        logger.error { "Expected to run ${it.key} but no events received" }
+                        false
+                    }
+
+                    is TestState.Failed -> false
+                    is TestState.Failing -> {
+                        logger.error { "Expected to run ${it.key} more, but no terminal events received" }
+                        false
+                    }
+
+                    is TestState.Passed -> true
+                    is TestState.Passing -> {
+                        logger.error { "Expected to run ${it.key} more, but no terminal events received" }
+                        //The test is passing but the execution mode might require all the 
+                        //runs to pass before considering this an actual pass 
+                        executionStrategy.mode == ExecutionMode.ANY_SUCCESS
+                    }
+                }
+            }.reduce { a, b -> a && b }
+        }
+    }
+
+    fun progress(): Float {
+        var done = 0
+        var total = 0
+
+        tests.values.forEach {
+            val state = it.state
+            when (state) {
```

Comment:
Today I learned you could do this when the IDE recommended it 🙂
```suggestion
            when (val state = it.state) {
```

Reply:
You can also ignore IDE and language features for the sake of simplicity and readability ;)
I prefer to define variables separately from the branching logic

Reply:
Agreed.